File: algorithms/solutions/java/_0101_SymmetricTree.java
/*
 * @lc app=leetcode.cn id=101 lang=java
 *
 * [101] Symmetric Tree
 *
 * https://leetcode.cn/problems/symmetric-tree/description/
 *
 * algorithms
 * Easy (60.15%)
 * Likes:    2700
 * Dislikes: 0
 * Total Accepted:    1M
 * Total Submissions: 1.7M
 * Testcase Example:  '[1,2,2,3,4,4,3]'
 *
 * Given the root of a binary tree, check whether it is a mirror of itself
 * (i.e., symmetric around its center).
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: root = [1,2,2,3,4,4,3]
 * Output: true
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: root = [1,2,2,null,3,null,3]
 * Output: false
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * The number of nodes in the tree is in the range [1, 1000].
 * -100 <= Node.val <= 100
 * 
 * 
 * 
 * Follow up: Could you solve it both recursively and iteratively?
 */

public class _0101_SymmetricTree {
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
// @lc code=start
    class Solution {
        public boolean isSymmetric(TreeNode root) {
            if (root == null) return true;
            return helper(root.left, root.right);
        }

        private boolean helper(TreeNode left, TreeNode right){
            if (left == null && right == null) return true;
            if (left == null || right == null) return false;
            return left.val == right.val && helper(left.left, right.right) && helper(left.right, right.left);
        }
    }
// @lc code=end
}

File: algorithms/solutions/python3/0101_symmetric_tree.py
#
# @lc app=leetcode.cn id=101 lang=python3
#
# [101] Symmetric Tree
#
# https://leetcode.cn/problems/symmetric-tree/description/
#

# @lc code=start
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        
        def helper(left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
            if not left and not right:
                return True
            if not left:
                return False
            if not right:
                return False
            return left.val == right.val and helper(left.left, right.right) and helper(left.right, right.left)
        
        return helper(root.left, root.right)
            
        
# @lc code=end

